### PR TITLE
Add muxer compliance test suite and refactor tests to not depend on `libp2p-swarm`

### PR DESCRIPTION
**Pull request by thomaseizinger**
# Description

The first part is likely to not be controversial.
Migrating away from `libp2p-swarm` avoids circular dependencies across our workspace. `libp2p-quic` is a transport and I'd argue that if the implementation only depends on `libp2p-core`, we should also test it outside of a `Swarm` context.

For example, this brings up the question on what to do with the `wrong_peer_id` test. Previously, this was testing functionality that is in `libp2p-swarm`.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/quic/tests/stream_compliance.rs
Comment (mxinden) [inline transports/quic/tests/stream_compliance.rs:15]: This is not QUIC protocol specific, but implementation specific, right?

File: transports/quic/tests/stream_compliance.rs
Comment (mxinden) [inline transports/quic/tests/stream_compliance.rs:8]: Wonderful to see these tests reused here. Great work @thomaseizinger!

**Review by mxinden — APPROVED** [review]
This looks good to me. Thanks @thomaseizinger.

Great to see the muxer tests used here.

I suggest waiting for @elenaf9's approval.

**Comment by kpp** [comment]
> For example, this brings up the question on what to do with the wrong_peer_id test. Previously, this was testing functionality that is in libp2p-swarm.

There was an intention to test TLS handshakes: you can't setup a connection if you dial with one PeerID and you get another in response.

File: transports/quic/tests/stream_compliance.rs
Comment (kpp) [reply transports/quic/tests/stream_compliance.rs:15]: @thomaseizinger isn't this one can be tricked with sending PING and waiting for PONG?

File: transports/quic/tests/smoke.rs
Comment (kpp) [inline transports/quic/tests/smoke.rs:52]: Seems like you are stuck to async-std tests? Why then there is a tokio test?

**Comment by thomaseizinger** [comment]
> > For example, this brings up the question on what to do with the wrong_peer_id test. Previously, this was testing functionality that is in libp2p-swarm.
> 
> There was an intention to test TLS handshakes: you can't setup a connection if you dial with one PeerID and you get another in response.

Fair enough. But to actually test this on the TLS level, we need to pass in the expected peer ID as part of creating the config:

https://github.com/libp2p/rust-libp2p/blob/f9b4af3d9d5b12b20756e496349b0866baa862da/transports/tls/src/lib.rs#L44

But we currently pass `None` here:

https://github.com/kpp/rust-libp2p/blob/58c3f67dc0d49890a407943c456b4e8ecee606a1/transports/quic/src/endpoint.rs#L82

The code will have to undergo some refactoring where we don't reuse the same config for all connections so we can pass the expected `PeerId` in here.

I'd suggest we do this as a follow-up. The current test only validates logic that is present in `libp2p-swarm` so we are not losing anything but not running this test atm.

File: transports/quic/tests/smoke.rs
Comment (thomaseizinger) [reply transports/quic/tests/smoke.rs:52]: What do you mean stuck to? We do test both providers in here.

File: transports/quic/tests/stream_compliance.rs
Comment (thomaseizinger) [reply transports/quic/tests/stream_compliance.rs:15]: Not quite, this one tests that we can read from a stream that we have never written to. Both `yamux` and QUIC seem to fail at that but I think in both cases, it is an implementation problem.

In reality, this isn't much of an issue in `libp2p` today because we always run multistream-select on the stream before we use it for the protocol. However, this also affects the TTFB metric and it would be great if we at some point have a muxer or a mechanism where we can open a stream that is already dedicated to a protocol. As such, I think these tests are worth having and we should fix the implementation eventually so they pass this test. The identify protocol f.e. opens a stream and requires the other party to just send data without us ever writing to the stream.

File: transports/quic/tests/smoke.rs
Comment (kpp) [reply transports/quic/tests/smoke.rs:52]: dial_failure is tested with async-std only. so does ipv4_dial_ipv6, wrong_peerid, tcp_and_quic, concurrent_connections_and_streams_async_std. endpoint_reuse is tested with tokio only.

Comment (thomaseizinger) [reply transports/quic/tests/smoke.rs:52]: I didn't really give it much thought to be honest so if any test is different from what it was before, it is an accident.

I can review again to retain which test uses which runtime if you want. I don't think it is particularly important as long as we have a smoke test for both!

Comment (thomaseizinger) [reply transports/quic/tests/smoke.rs:52]: I just went over it again and every test is still run with the same executor as before. `endpoint_reuse` also uses `tokio` on the current `libp2p-quic` branch so this PR doesn't change any of that!

I'd prefer if we can sort this in a follow-up :)

File: transports/quic/tests/stream_compliance.rs
Comment (mxinden) [reply transports/quic/tests/stream_compliance.rs:15]: > However, this also affects the TTFB metric and it would be great if we at some point have a muxer or a mechanism where we can open a stream that is already dedicated to a protocol.

Note that with [`V1Lazy`](https://docs.rs/multistream-select/0.10.2/multistream_select/enum.Version.html#variant.V1Lazy) and only a single stream protocol, multistream-select does not have an impact on TTFB.

That said, agreed. We should fix this.

Comment (thomaseizinger) [reply transports/quic/tests/stream_compliance.rs:15]: > > However, this also affects the TTFB metric and it would be great if we at some point have a muxer or a mechanism where we can open a stream that is already dedicated to a protocol.
> 
> Note that with [`V1Lazy`](https://docs.rs/multistream-select/0.10.2/multistream_select/enum.Version.html#variant.V1Lazy) and only a single stream protocol, multistream-select does not have an impact on TTFB.
> 
> That said, agreed. We should fix this.

It does for the listener right? In this particular testcase, the dialer doesn't write _any_ bytes to the stream and expects the listener to answer with the payload of a particular protocol right away.

File: transports/quic/tests/smoke.rs
Comment (elenaf9) [reply transports/quic/tests/smoke.rs:52]: The executor choice for most tests here is arbitrary, i.e. some use `async-std` and others `tokio` just because there is no reason to favor one over the other in our tests.  `concurrent_connections_and_streams` and the basic `smoke` test I tested with both runtimes to guarantee that the fundamental functionality works with both rumtimes.
Don't feel strongly about this.

File: transports/quic/tests/smoke.rs
Comment (elenaf9) [inline transports/quic/tests/smoke.rs]: Removing this test is fine by me.

File: transports/quic/tests/smoke.rs
Comment (elenaf9) [inline transports/quic/tests/smoke.rs:164]: This test hangs forever when I run it locally.

File: transports/quic/tests/stream_compliance.rs
Comment (elenaf9) [reply transports/quic/tests/stream_compliance.rs:15]: If we create a new stream, the other side won't be notified of the new stream until we actually start sending data. See quinn-rs/quinn#1425. Imo this makes sense since the initiator has to send some data on the stream first anyway so that the remote knows which application layer protocol belongs to (like it is done today with multistream-select, as @thomaseizinger already described above). 

> if we at some point have a muxer or a mechanism where we can open a stream that is already dedicated to a protocol.

Given that such a muxer is not planned for the near future I don't think this is something we have to fix or test now. 
The idea that one side simply sends an empty STREAM frame and with this triggers the remote to start sending a bunch of data on that stream generally feels a bit weird. Happy to be convinced otherwise :)
But I think that's out of scope for this PR, thus I'd be in favor of removing this test.

**Review by elenaf9 — CHANGES_REQUESTED** [review]
Sorry for the delay. Thanks @thomaseizinger!

The `concurrent_connections_and_streams_async_std` test runs forever when I run it locally. Any idea why that could be the case @thomaseizinger?

File: transports/quic/tests/smoke.rs
Comment (thomaseizinger) [reply transports/quic/tests/smoke.rs:164]: Are you sure it hangs forever and doesn't just take long? I had this test take up to 90s some times locally although that didn't happen recently.

File: transports/quic/tests/stream_compliance.rs
Comment (thomaseizinger) [reply transports/quic/tests/stream_compliance.rs:15]: Happy to remove it from the test suite. Will do in a separate PR!

Comment (elenaf9) [reply transports/quic/tests/stream_compliance.rs:15]: Sorry, I probably wasn't clear with what I meant. I don't think the test needs to be removed completely from the test-suite, since we may use it in the future to test a muxer that implements what you described above. I was just suggestion to remove this test within the QUIC transport since we will most likely not support it here.

File: transports/quic/tests/smoke.rs
Comment (elenaf9) [reply transports/quic/tests/smoke.rs:164]: Found the issue by enabling the `nocapture` flag: The initial handshake timed out, which caused the spawned tasks to panic in l. 279 and l. 260. Because this happens in a separate task the test itself did not panic. Setting the handshake timeout to 5s fixes it.

It would be great if the test could panic if one of the spawned tasks panics. Maybe simply by having something like a oneshot channel for each task so we get notified if it dropped?

File: transports/quic/tests/stream_compliance.rs
Comment (thomaseizinger) [reply transports/quic/tests/stream_compliance.rs:15]: I don't think there is any point in having a test in the test suite that we don't end up running for every muxer.

File: transports/quic/tests/smoke.rs
Comment (thomaseizinger) [reply transports/quic/tests/smoke.rs:164]: If we don't spawn separate tasks, then the test is not actually concurrent!

I added a timeout to the final `block_on` which should trigger if this happens again.

File: transports/quic/tests/smoke.rs
Comment (thomaseizinger) [reply transports/quic/tests/smoke.rs]: I changed the comment to say it should be re-written once we do the refactoring.

File: transports/quic/tests/smoke.rs
Comment (elenaf9) [inline transports/quic/tests/smoke.rs]:
```suggestion
```
As mentioned in https://github.com/kpp/rust-libp2p/pull/27#discussion_r1018919787, timeout of 1s is too short when I run it locally. It triggers a `HandshakeTimeout` for the `concurrent_connections_and_streams_async_std` test. Tests pass for me with the default value, which is 5s.

Comment (thomaseizinger) [reply transports/quic/tests/smoke.rs]: Changed it to use the default config!

**Comment by thomaseizinger** [comment]
I don't know why GitHub is stupid and doesn't let me re-request a review from both of you @elenaf9 @kpp. In any case, I'd appreciate another review :)

**Review by elenaf9 — APPROVED** [review]
Thanks @thomaseizinger!

**Review by kpp — APPROVED** [review]

**Pull request merged**
